src/imitation/contrib/data_loader_wrappers.py

```diff
@@ -0,0 +1,211 @@
+        `VecEnv.step`, or `DataLoader.__iter__`.
+
+        `forwards` should be a pure function. In other words, any instance of
+        BidirectionalTransform returns the same outputs given the same inputs.
+
+        Args:
+            transitions_data: A Mapping roughly corresponding to an instance of
+                Transitions or TransitionsWithRew. It must have the keys "obs", "acts",
+                "next_obs", and "infos". Optionally it can include the "rews" key.
+
+                Concrete implementations of this method that transform "rews" must be
+                able to gracefully handle the case where "rews" is not provided.
+
+        Returns:
+            A dict with the same keys as `transitions_data`, but where some values
+            have been replaced.
+        """
+
+    @abc.abstractmethod
+    def forwards_obs(
+        self,
+        obs: Any,
+    ) -> Any:
+        """Defines a batched transformation from raw to transformed observations.
+
+        Used transform observations returned by `gym.Wrapper.reset()`.
+        """
+
+    @abc.abstractmethod
+    def backwards_act(self, transformed_acts: np.ndarray):
+        """Defines a (pseudo-)inverse transformation from transformed act to raw act.
+
+        `backwards_act` should be a pure function. In other words, any instance of
+        BidirectionalTransform returns the same outputs given the same inputs.
+
+        Args:
+            transformed_acts: Batched actions transformed from raw actions by
+                `forwards()`.
+
+        Returns:
+            A raw action. Not required to be the same raw action as that passed into
+            `forwards()` because `forwards()` might not be one-to-one / injective.
+        """
+
+
+class BidirectionalDataLoaderWrapper(DataLoaderWrapper[Mapping, dict]):
+    """DataLoaderWrapper that corresponds to an instance of BidirectionalTransform.
+
+    Primarily used by BidirectionalGymWrapper.
+    """
+
+    def __init__(
+        self,
+        data_loader: types.DataLoaderInterface,
+        transform: BidirectionalTransform,
+    ):
+        self.data_loader = data_loader
+        self.transform = transform
+        super().__init__(data_loader)
+
+    def __iter__(self) -> dict:
+        for trans in self.data_loader:
+            result = self.transform.forwards(trans)
+            result["acts"] = self.transform.backwards_act(trans["acts"])
+            yield result
+
+
+class DefinesDataLoaderWrapper(abc.ABC, Generic[S, T]):
+    """Concrete instances of this abstract mixin define a `DatasetWrapper`.
+
+    Intended for use as an abstract mix-in in a subclass of `gym.Wrapper` so that the
+    `gym.Wrapper` can be processed by `wrap_data_loader_with_env_wrappers`.
+    """
+
+    @abc.abstractmethod
+    def apply_dataset_wrapper(
+        self,
+        data_loader: types.DataLoaderInterface,
+    ) -> DataLoaderWrapper[S, T]:
```

Comment:
When would we want to do something other than applying a `BidirectionalTransform` here?

Reply:
Something like FrameStack, if we ever decide to support that. Leaving the API open to allowing non-BidirectionalTransform for now.

Reply:
To make sure I understand, `FrameStack` can't be supported in `BidirectionalTransform` because it needs to be stateful? This would also require that the data loader returns transitions in sequence, I guess.